### ISSUE TITLE
col_number unable to return int

### ISSUE THREAD

**Issue by momeara** [tidyverse/readr#807]
parsing with `col_number` always returns a `dbl` even when it would be better to return an `int`

```
library(readr)
read_tsv(
   "ID\nID01\nID02\nID03",
   col_types=cols(col_number()))
```
returns
```
 # A tibble: 3 x 1
           ID
        <dbl>
 1         1.
 2         2.
 3         3.
```
The column type for `ID` is `<dbl>`, but it should be `<int>`.

**Comment by batpigandme** [comment]
What if you use `col_integer()`?

**Comment by momeara** [comment]
Using `col_integer` fails to strip the non-numeric characters resulting in `NA` values:

```
library(readr)
read_tsv(
   "ID\nID01\nID02\nID03",
   col_types=cols(col_integer()))
```
returns
```
 Warning: 3 parsing failures.
 row col   expected actual         file
   1  ID an integer   ID01 literal data
   2  ID an integer   ID02 literal data
   3  ID an integer   ID03 literal data

 # A tibble: 3 x 1
      ID
   <int>
 1    NA
 2    NA
 3    NA
```

**Comment by batpigandme** [comment]
What does the actual tsv file for that look like? I tried to see if the issue had to do with the leading zero, but it works when I use `.tsv` and `.csv` files I've made up.
gist here: https://gist.github.com/batpigandme/7655ae3a534572bab7d6aee526275fb9
``` r
library(readr)
readr::read_csv("https://gist.githubusercontent.com/batpigandme/7655ae3a534572bab7d6aee526275fb9/raw/ca90b4f55c4bbe99aadb09200b371e62ee7bf97f/readrex.csv", col_types = cols(ID = col_integer(), 
                                                                                                                                                                             ID0 = col_number()))
#> # A tibble: 3 x 2
#>     ID0    ID
#>   <dbl> <int>
#> 1    1.     1
#> 2    2.     2
#> 3    3.     3
readr::read_tsv("https://gist.githubusercontent.com/batpigandme/7655ae3a534572bab7d6aee526275fb9/raw/8e2dfdee326b12443497035cdc7963d2d6028003/readrex.tsv", col_types = cols(ID = col_integer(), 
                                                                                                                                                                             ID0 = col_number()))
#> # A tibble: 3 x 2
#>     ID0    ID
#>   <dbl> <int>
#> 1    1.     1
#> 2    2.     2
#> 3    3.     3
readr::read_tsv("https://gist.githubusercontent.com/batpigandme/7655ae3a534572bab7d6aee526275fb9/raw/8e2dfdee326b12443497035cdc7963d2d6028003/readrex.tsv", col_types = cols(ID = col_integer(), 
                                                                                                                                                                             ID0 = col_integer()))
#> # A tibble: 3 x 2
#>     ID0    ID
#>   <int> <int>
#> 1     1     1
#> 2     2     2
#> 3     3     3
```
Created on 2018-03-05 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by momeara** [comment]
The `tsv` file is given as a string, but if it was read in from a file, the file would look like this:

```
ID
ID01
ID02
ID03
```

If the `0`'s are stripped then the results are the same:

```
read_tsv(
   "ID\nID1\nID2\nID3",
   col_types=cols(col_number()))
```
returns
```
 # A tibble: 3 x 1
      ID
   <dbl>
 1    1.
 2    2.
 3    3.
```

and 
```
read_tsv(
    "ID\nID1\nID2\nID3",   col_types=cols(col_integer()))
```
gives
```
 Warning: 3 parsing failures.
 row col   expected actual         file
   1  ID an integer    ID1 literal data
   2  ID an integer    ID2 literal data
   3  ID an integer    ID3 literal data

 # A tibble: 3 x 1
      ID
   <int>
 1    NA
 2    NA
 3    NA
```

**Comment by batpigandme** [comment]
OK, thanks. I think this at least suggests that there's something different in how it's parsing the character-prefixed data… I don't know the answer, was just trying to help whittle things down.

**Comment by cderv** [comment]
Only `parse_number` can drops non numeric character to extract numbers in string based on the local decimal mark and grouping mark. It is why `parse_integer` fails. 
I think by design, `parse_number` results in numeric. I don't know if it would be possible to assume integer if there is no decimal mark. I think it could be costly to guess that and converting to integer after parsing seems not so painful. Everything happens in the C++ code in `parseNumber`, still not so easy to read for me.

So, I am not sure this is bug,  but more a design choice or a limitation that would worth documenting more precisely in `help("parse_number')`. I could do a PR but I'll wait for confirmation on how `parse_number` is intended to work.

**Comment by jimhester** [comment]
As stated parse_number always returning R numeric vectors is by design.

**Comment by jimhester** [comment]
Fixed by 9a92f23d74
